digital_land/update.py

```diff
@@ -66,33 +70,151 @@ def has_collected_resource(log_item):
+    endpoint_entries = endpoint_register.entries
+    if resource_entry["endpoint"] in endpoint_register.record:
+        existing_idx = endpoint_register.record[resource_entry["endpoint"]][0]
+        if endpoint_entries[existing_idx].item["end-date"]:
+            # Entry already exists with an end-date
+            print(
+                "WARNING: endpoint end-date {} found for URL {}".format(
+                    endpoint_entries[existing_idx].item["end-date"],
+                    resource_entry["endpoint-url"],
+                )
+            )
+        else:
+            # No op if active entry already exists
+            print(
+                "Active endpoint already exists for URL {}".format(
+                    resource_entry["endpoint-url"]
+                )
+            )
+            return
+
+    endpoint_entry = {
+        "endpoint": resource_entry.get("endpoint"),
+        "endpoint-url": resource_entry.get("endpoint-url"),
+        "entry-date": date.today().strftime("%Y-%m-%d"),
+        "end-date": datetime.strptime(resource_entry["end-date"], "%Y-%m-%d").date()
+        if "end-date" in resource_entry
+        else "",
+    }
+
+    # If empty start-date set by user then no date should be set. Otherwise set user-specified date or use current date
+    if "start-date" in resource_entry:
+        if resource_entry["start-date"]:
+            endpoint_entry["start-date"] = datetime.strptime(
+                resource_entry["start-date"], "%Y-%m-%d"
+            ).date()
+    else:
+        endpoint_entry["start-date"] = date.today().strftime("%Y-%m-%d")
+
+    endpoint_item = Item(endpoint_entry)
+    endpoint_register.add(endpoint_item)
+
+
+def add_new_source(resource_entry, source_register):
+    source_entries = source_register.entries
+    if resource_entry["endpoint"] in source_register.record:
+        for idx in source_register.record[resource_entry["endpoint"]]:
+            if (
+                resource_entry["organisation"]
+                == source_entries[idx].item["organisation"]
+            ):
+                if source_entries[idx].item["end-date"]:
+                    # Entry already exists with an end-date
+                    print(
+                        "WARNING: source end-date {} found for URL {}".format(
+                            source_entries[idx].item["end-date"],
+                            resource_entry["endpoint-url"],
+                        )
+                    )
+                    break
+                else:
+                    # No op if active entry already exists
+                    print(
+                        "Active source entry already exists for organisation {} and URL {}".format(
+                            resource_entry["organisation"],
+                            resource_entry["endpoint-url"],
+                        )
+                    )
+                    return
+
+    source_entry = {
+        "collection": resource_entry.get("pipeline", ""),
```

Comment:
you could use a `defaultdict(str)` here in place of the dict for resource_entry, it would avoid the use of `.get` all over the place, because `""` would be the default for any missing key.

Reply:
Updated